### PR TITLE
caching tutorial

### PR DESCRIPTION
**Pull request by jardakotesovec**
Just finished first version of caching tutorial. Any feedback (including english corrections) appreciated.

Thanks!

### CODE REVIEW

**Comment by nlf** [comment]
Wow! At first glance, this is awesome! Great work!

When I get some time I'll review it more closely and we'll get it merged. Thank you for this!

File: lib/tutorials/caching.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
+## Client side caching
+
+Http protocol provides several mechanisms how to cache resources in browser via http headers. To decide which headers are suitable for your use case check Google developers [guide](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching) or other [resources](https://www.google.com/search?q=HTTP+caching). This tutorial provides overview how to use these techniques in hapi.
```
Comment (arb) [inline lib/tutorials/caching.md]: > mechanisms how to cache resources in browser via http headers
maybe change to "different Http headers to control how browsers cache resources."

File: lib/tutorials/caching.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+### Cache-Control
+
+`Cache-Control` header tells browser and any intermediate cache if resource is cacheable and for how long. For example `Cache-Control:max-age=30, must-revalidate, private` means that browser can cache resource for 30 seconds and `private` means that it should not be cached by intermediate caches, only by browser. `must-revalidate` means that once it expires it has to request the resource again.
```
Comment (arb) [inline lib/tutorials/caching.md]: "The `Cache-Contol` header"

I think you need to put "the" before the word "browser" in this whole paragraph.

File: lib/tutorials/caching.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      }
+   },
+   config: {
+      cache: {
+         expiresIn: 30 * 1000,
+         privacy: 'private'
+      }
+   }
+});
+```
```
Comment (arb) [inline lib/tutorials/caching.md]: I generally name my code examples to I can refer to them easier; IE **Listing 1**.

File: lib/tutorials/caching.md (hunk 1 above)
Comment (nlf) [inline lib/tutorials/caching.md]: In addition to that suggestion, HTTP should always be in all caps

File: lib/tutorials/caching.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+Example also illustrates that `expiresIn` value can be overridden with `ttl(msec)` function using [response object](http://hapijs.com/api#response-object) interface.
+
+See [route-options](http://hapijs.com/api#route-options) for more options.
+### Last-Modified
+
+In some cases server can provide information when resources was last modified. Good example is [inert](https://github.com/hapijs/inert) module, which is responsible for serving static files in hapi (`file` and `directory` handler). Last modification timestamp is read for every requested file and send as `Last-Modified` header.
```
Comment (arb) [inline lib/tutorials/caching.md]: - "The server"
- "A good example"

Maybe reward the sentence about inert:
When using either the `file` or `directory` route configuration option, the underlying module, inert.... bla bla bla.

File: lib/tutorials/caching.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+When `Last-Modified` is set on response hapi compares it with `If-Modified-Since` coming from client to decide if should return `304`.
```
Comment (arb) [inline lib/tutorials/caching.md]: "if the response status code should be 304."

File: lib/tutorials/caching.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+
+Assuming `lastModified` is Date object you can set this header via [response object](http://hapijs.com/api#response-object) interface.
```
Comment (arb) [inline lib/tutorials/caching.md]: Is this something the average person will need to know? Seems like something we could cut.

File: lib/tutorials/caching.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+```
+   reply(result).header('Last-Modified', lastModified.toUTCString());
+```
+
+There is one more example using `Last-Modified` in [last section](#client-and-server-caching).
+
+### ETag
+
+ETag is alternative to `last-modified` where server provides token (usually checksum of resource) instead of last modified timestamp. Browser in next request sends this ETag in `If-None-Match` header. Server compares it with new `ETag` and return `304` if it's the same.
```
Comment (arb) [inline lib/tutorials/caching.md]: The "ETag" header is...

File: lib/tutorials/caching.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+```
+   reply(result).etag('xxxxxxxxx');
+```
+
+Check `options` parameter for `etag` function on [api](http://hapijs.com/api#response-object) page for more details.
+
+
+## Server side caching
+
+hapi provides powerful server side caching via [catbox](https://www.github.com/hapijs/catbox) and makes using cache very convenient. This tutorial help you to understand how catbox is utilized inside hapi and how you can use it. Catbox also can be used out of the hapi, but that is not covered here.
```
Comment (arb) [inline lib/tutorials/caching.md]: maybe how you can "leverage" it. Sounds fancier :grinning:

Comment (arb) [inline lib/tutorials/caching.md]: Maybe remove the sentence about using catbox outside of hapi. Doesn't really seem helpful here.

File: lib/tutorials/caching.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+Catbox has two interfaces.
+
+
+### Client
+[Client](https://github.com/hapijs/catbox#client) is low-level interface, which basically allows you set/get key-value pairs. Its initialized with one of the available adapters ([Memory](https://github.com/hapijs/catbox-memory), [Redis](https://github.com/hapijs/catbox-redis), [mongoDB](https://github.com/hapijs/catbox-mongodb), [Memcached](https://github.com/hapijs/catbox-memcached), [Riak](https://github.com/DanielBarnes/catbox-riak), [Amazon S3](https://github.com/fhemberger/catbox-s3)).
```
Comment (arb) [inline lib/tutorials/caching.md]: Should we include Amazon S3? It's technically not an "official" plugin.

File: lib/tutorials/caching.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+       }
+    ]
+});
+
+server.connection({
+    port: 8000
+});
+```
+
+In this example two [clients](https://github.com/hapijs/catbox#client) were defined: `'mongoCache'`, `'redisCache'` and also default one without name with [memory](https://github.com/hapijs/catbox-memory) adapter. You can replace default client by omitting `name` property. `partition` property tells adapter how cache should be named (`'catbox'` by default). In case of [mongoDB](http://www.mongodb.org/) it is database name and in case of [redis](http://redis.io/) it is used as key prefix.
```
Comment (arb) [inline lib/tutorials/caching.md]: Maybe reword to something like:
In Listing 2, we've defined two catbox clients; mongoCache and redisCache. Including the default memory cache created by hapi, there are three available cache clients.

File: lib/tutorials/caching.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    }
+});
+
+server.route({
+    path: '/add/{a}/{b}',
+    method: 'GET',
+    handler: function (request, reply) {
+
+        var id = request.params.a + ':' + request.params.b;
+        sumCache.get({id: id, a: request.params.a, b: request.params.b}, function (err, result) {
```
Comment (arb) [inline lib/tutorials/caching.md]: Don't you only need the `id` key here in this object? I don't think `a` and `b` do anything?

File: lib/tutorials/caching.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+server.connection({
+    port: 8000
+});
+```
+
+In this example two [clients](https://github.com/hapijs/catbox#client) were defined: `'mongoCache'`, `'redisCache'` and also default one without name with [memory](https://github.com/hapijs/catbox-memory) adapter. You can replace default client by omitting `name` property. `partition` property tells adapter how cache should be named (`'catbox'` by default). In case of [mongoDB](http://www.mongodb.org/) it is database name and in case of [redis](http://redis.io/) it is used as key prefix.
+
+### Policy
+[Policy](https://github.com/hapijs/catbox#policy) is high-level interface that you will be using. Lets pretend we are dealing with something more complicated than adding two numbers and we want to cache the results. [server.cache](http://hapijs.com/api#servercacheoptions) method creates new [policy](https://github.com/hapijs/catbox#policy), which is than used in handler. `id` object has mandatory key `id`, which is a unique item identifier string.
```
Comment (arb) [inline lib/tutorials/caching.md]: Maybe remove the last sentence and create a new paragraph explaining `sumCache.get` below the code. 

It might be worth explaining what the `generateFunc` does (like why the signature is the way it is) as well.

File: lib/tutorials/caching.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+In addition to **partitions** there are **segments** that allows to isolate caches within one [client](https://github.com/hapijs/catbox#client). If you want to cache results from two different methods, you usually don't want mix up the results. In [mongoDB](http://www.mongodb.org/) `segment` represents collection and in [redis](http://redis.io/) it's additional prefix to `partition` prefix. Default value for `segment` is provided when [server.cache](http://hapijs.com/api#servercacheoptions) is called inside plugin - `'!pluginName'` or used via [server methods](http://hapijs.com/tutorials/server-methods) - `'#methodName'`. If you have use case for multiple policies sharing one segment there is [shared](http://hapijs.com/api#servercacheoptions) option.
+
+### Server methods
+But it can get better than that! In 95% cases you will use [server methods](http://hapijs.com/tutorials/server-methods) for caching purposes, because it reduces boilerplate to minimum. Lets do the same with server methods:
+
+```javascript
+server.method('sum', add, { cache: { cache: 'mongoCache', expiresIn: 30 * 1000 } });
```
Comment (arb) [inline lib/tutorials/caching.md]: For completeness, maybe include the `add` function here as well?

File: lib/tutorials/caching.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+
+In this case we use `cached.stored` timestamp to set `last-modified` header.
+
+```javascript
+server.route({
+    path: '/add/{a}/{b}',
+    method: 'GET',
+    handler: function (request, reply) {
+
+        server.methods.sum(request.params.a, request.params.b, function (err, result, cached, report) {
```
Comment (arb) [inline lib/tutorials/caching.md]: You could probably add these two parameters to the previous example so you didn't need to create a whole new one.

File: lib/tutorials/caching.md (hunk 6 above)
Comment (jardakotesovec) [inline lib/tutorials/caching.md]: Do you mean whole Last-Modified section? There is good example how to use Last-Modified with server side caching that can be pretty useful (in last section).

File: lib/tutorials/caching.md (hunk 11 above)
Comment (jardakotesovec) [inline lib/tutorials/caching.md]: You have to pass the params somehow (thats how server-methods do it) to get function so `generateFunc` can access them and can to do something with it.

**Comment by arb** [comment]
@arb to make second English pass.
